### PR TITLE
[Reland] add DisableTorchFunction that matches DisableTorchDispatch (#88219)

### LINKED ISSUE #87990: Dispatch mode and torch function disable are inconsistent
### 🐛 Describe the bug

`no_dispatch` disables both mode and subclass (because it excludes the Python dispatch key, which controls for both)

`_C.DisableTorchFunction` only disables subclass (this behavior is explicitly tested for in `test_mode_notimplemented_loop`
and `test_disable_subclass_not_mode`)

Probably easiest fix is to rename DisableTorchFunction to DisableTorchFunctionSubclass and then add a mode disabler.

### Versions

master

### PR DESCRIPTION
**Pull request by samdow**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #92012

Reland of #88219

Closes #87990. This implements a new disable guard that matches DisableTorchDispatch (disables all subclasses and modes)

### CODE REVIEW

**Comment by samdow** [comment]
Now that #89221 is in, this can land too

File: test/test_public_bindings.py [hunk 1]
```diff
@@ -99,6 +99,7 @@ def test_no_new_bindings(self):
             "device",
             "DeviceObjType",
             "DictType",
+            "DisableTorchFunction",
```
Comment (albanD) [inline test/test_public_bindings.py:102]: Do we really want this to be in `torch.` ? Not `torch.overrides` or another torch_function related namespace?

**Review by albanD — COMMENTED** [review]
Code change sounds good, just a question about location of the new API

**Comment by ezyang** [comment]
Deferring to alban

File: test/test_public_bindings.py (hunk 1 above)
Comment (samdow) [reply test/test_public_bindings.py:102]: So at minimum I argue that DisableTorchFunction and DisableTorchFunctionSubclass should be in the same place (either `torch.overrides` or `torch.`). I can put up an issue but given how hard it was to land #89221 which was just a rename I'm not confident that I can move both easily

Comment (albanD) [reply test/test_public_bindings.py:102]: Ok, let's do it this way then...

**Review by albanD — APPROVED** [review]

**Pull request closed without merging**